### PR TITLE
Fix missing extent cache invalidation

### PR DESCRIPTION
**Pull request by brisvag**
# References and relevant issues
Started on zulip: https://napari.zulipchat.com/#narrow/stream/212875-general/topic/Layer.2Erefresh.28.29.20doesn't.20reset.20dims.20range

# Description
The issue at the core is the fact that we cache extents, and are unable to invalidate this cache when layer data is changed inplace (since we have no events associated with the array inside).

Example failing code:

```py
import napari
import numpy as np
viewer = napari.Viewer()
points = np.array([[0, 1, 2], [1, 2, 3], [2, 3, 4], [3, 4, 5]])
actual_slices = np.array([4, 6, 8, 10])

layer = viewer.add_points(points)
viewer.layers['points'].data[:,0] = actual_slices
```

This should result in dims going from 0 to 6, with alternating empty slices. Instead, we get 0-3, with only empty slices, cause the points layer extents are outdated.

Adding the following lines will fix the issue:

```py
layer._clear_extents_and_refresh()  # invalidates cached extents
viewer._on_layers_change()  # recomputes dims extents
```

For now, I only switched from `refresh()` to `_clear_extents_and_refresh()` where we do have knowledge of a change in data. However, to actually fix the above code, we need to somehow either:

1. always invalidate the cache on `refresh()` for non array-like layers (might be a big hit on performance cause we refresh in a *lot* of places that don't really need this
2. expose a "*hard refresh*" method
3. change `refresh` to always invalidate cahce, but expose a "*soft refresh*" that does not, and use that throughout napari every time we don't need cache invalidation

I think 3. is the best, but it's also a bit more invasive and I'm not sure how many people use refresh and rely on it being performant outside of napari core.

### CODE REVIEW

**Comment by Czaki** [comment]
I do not think, that we should support modification data in place. 

I suggest that people should use 

`layer.data = layer.data` to refresh. 

All other solution is high risk for performance problem or inconsistency. I do not feel like current code of layers will allow for this. 

Maby after big refactor.

**Comment by brisvag** [comment]
> I do not think, that we should support modification data in place.

I'm not saying we should be automatically firing events (I agree that's too much work we need to inject into arrays and risk having huge performance issues), but I definitely think we should *support* setting data inplace (and not straight up explode or fail) as long as there's a manual way of letting napari now "Data has changed!".

But yeah, we can also encourage people to use `layer.data = data`, but that's also a bigger performance hit than needed, in many cases.

**Comment by Czaki** [comment]
> But yeah, we can also encourage people to use `layer.data = data`, but that's also a bigger performance hit than needed, in many cases.

Why is it a bigger performance hit? If you allow modifying in place. Then how it will differ from setting, a new data?

**Comment by brisvag** [comment]
It's a bigger performance hit than just doing `layer.refresh()`, because `layer.refresh()` does not invalidate the extent caches. There are many cases where people use `refresh()` where they don't need to recompute extents, and if they do so as part of a rapid-firing event it might be significantly slower.

**Comment by Czaki** [comment]
I know, that refresh is less expensive (it is still too expensive - for example trigger recomputation of thumbnail). 

But I do not see how modify data in place could be less expensive than assign data.

**Comment by psobolewskiPhD** [comment]
I kinda like the idea of the hard refresh that the user can do? 
Basically take the layer.data and treat like I just added it as a new layer?
That's what I expected .refresh() to do frankly.
I don't think we need to support in-place edits with auto-magic, but making it so someone could do it would be nice.
The specific case that brought up this issue was someone who gets a layer generated by a plugin, but wants to fix the ordering of the slices. So doing it in-place is the easiest, because they don't have any/easy access to the original array that was passed as data for the layer.

**Comment by jni** [comment]
>  That's what I expected .refresh() to do frankly.

👆 I agree wholeheartedly with this. If something is *not* refreshed by layer.refresh, I consider that a bug. Obviously, in many places we need partial refresh — we can add support for that in other ways. e.g. `layer.refresh(thumbnail=False, extent=False)`?

**Comment by Czaki** [comment]
> 👆 I agree wholeheartedly with this. If something is _not_ refreshed by layer.refresh, I consider that a bug. Obviously, in many places we need partial refresh — we can add support for that in other ways. e.g. `layer.refresh(thumbnail=False, extent=False)`?

You forget about colormap etc. 
The general problem is that refresh is used in multiple places and even now, provides us significant negative impact on performance of layers.

**Comment by jni** [comment]
>  The general problem is that refresh is used in multiple places and even now, provides us significant negative impact on performance of layers.

of course. That's a problem with our programming style, though. It does not change the point that as a user, I would definitely expect `refresh()` to actually refresh. It should be the last resort for forcing the viewer to sync changes.

Anyway, I wouldn't mind some middle ground, where all the attributes that can be refreshed are false by default, but we add `refresh_all()` that sets them all to true. Then we can audit our calls to `refresh` and set only necessary elements to True.

**Comment by brisvag** [comment]
What about something like this in the last commit?

Then I would have to change every occurence of `layer.refresh()` to use `False` for the unneeded parts.

**Comment by psobolewskiPhD** [comment]
Sorry to be so late to this.
I can appreciate not wanting to support/handle/encourge in place modification of data.
In the "trigger case" the user had a layer generated by a plugin, so they didn't have easy access to the "original" data. I warned that this wasn't a good way to do it!
I guess the solution should have been:
```
current_data = viewer.layers['the layer'].data
current_data[;,0] = actual_data
viewer.layers['the layer'].data = current_data
```

Still would be nice to do a "no seriously refresh everything like for real"

**Comment by jni** [comment]
> What about something like this in the last commit?
> 
> Then I would have to change every occurence of layer.refresh() to use False for the unneeded parts.

Personally I really like this approach!

**Comment by brisvag** [comment]
Let me know what you think. I went through all the refresh calls and updated it accordingly... it's a bit tricky to understand the control flow of our refreshes sometimes, so it would be great to get other eyes on it. I also expect some tests to fail and let me know where I messed up :P

I also left some TODOs where I wasn't sure what to do.

**Comment by brisvag** [comment]
I don't understand this mypy error. It points to clearly wrong lines, so I don't know how to debug it.

File: napari/layers/base/base.py
Comment (Czaki) [inline napari/layers/base/base.py]:
```suggestion
    def refresh(
        self,
        event: Optional[Event] = None,
        *,
        thumbnail: bool = True,
        data_displayed: bool = True,
        highlight: bool = True,
        extent: bool = True,
        force: bool = False,
    ) -> None:
```

File: napari/layers/base/base.py
Comment (Czaki) [inline napari/layers/base/base.py]:
```suggestion
    def _refresh_sync(
        self,
        *,
        thumbnail: bool = False,
        data_displayed: bool = False,
        highlight: bool = False,
        extent: bool = False,
        force: bool = False,
    ) -> None:
```

**Review by Czaki — COMMENTED** [review]
This should fix mypy error. 
If you commit it, please restart benchmarks

**Comment by brisvag** [comment]
So if I understand correclty, only 1 benchmark got worse, and it's a full refresh (which makes sense, cause now it's also refreshing the extent). I guess that's ok?

**Comment by Czaki** [comment]
> So if I understand correclty, only 1 benchmark got worse, and it's a full refresh (which makes sense, cause now it's also refreshing the extent). I guess that's ok?

I think yes.

File: napari/layers/base/base.py
Comment (Czaki) [inline napari/layers/base/base.py]:
```suggestion
        if not (self.visible or force):
            return
        if extent:
            self._clear_extent()
            self._clear_extent_augmented()
        if data_displayed:
            self.set_view_slice()
            self.events.set_data()
        if thumbnail:
            self._update_thumbnail()
        if highlight:
            self._set_highlight(force=True)
```

**Review by Czaki — APPROVED** [review]
I do not like this solution, but it looks like any better solution will require to refactor layers.

**Comment by brisvag** [comment]
> I do not like this solution, but it looks like any better solution will require to refactor layers.

yeah, this whole PR showed me how weird and entangled this whole process is... we really need to decouple render state from layers :/

**Review by brisvag — COMMENTED** [review]
I have a few TODOs in there, it would be good to get some feedback on them and address them before merging!

**Comment by jni** [comment]
Discussed in community meeting: @Czaki pointed out (and I agree) that cache invalidation is hard 😂 and we would probably benefit with longer testing on main, so we have decided to move it to the next milestone.

**Comment by jni** [comment]
(with an idea to merge immediately after the 0.5.1 release)

**Comment by brisvag** [comment]
@jni we didn't merge immediately after 0.5.1.... shoudl we still get this in?

**Comment by Czaki** [comment]
> @jni we didn't merge immediately after 0.5.1.... shoudl we still get this in?

I think that you should merge this just after 0.5.2 release

**Comment by jni** [comment]
Ach. I'll try to remember... 😕

**Comment by brisvag** [comment]
@jni reminder since you're in release mode and I thought about this!

**Comment by jni** [comment]
haha I also thought about this (especially because I think #7146 should also get merged quickly so we can all play with it on main). But your warning comes a few hours too early, and is thus useless! 😂 I have set a reminder on my watch to merge this in the morning though. 😃

**Review by Czaki — APPROVED** [review]
Let's play with this

**Pull request merged**
